segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -175,7 +176,7 @@ public void readSegment(ReadSegment readSegment) {
                     handleReadResult(readSegment, readResult);
                     readStreamSegment.reportSuccessEvent(timer.getElapsed());
                 })
-                .exceptionally(ex -> handleException(readSegment.getOffset(), segment, "Read segment", ex));
+                .exceptionally(ex -> handleException(readSegment.getOffset(), segment, "Read segment", wrapCancellationException(ex)));
```

Comment:
I'm not sure this is the right solution. At the very least segemntStore's java doc should specify what occurs when TIMEOUT is reached. But in general it might be better for it to return the ReadResult such as it is without waiting further.

Reply:
My initial reaction to the PR changes was similar. I also felt that perhaps the exception out of the segment store was not being sufficiently precise, but the alternative I have given was to wrap it in PRP. I assumed that cancellation exception for a read out of the segment store can only be due to a timeout, but perhaps my assumption is incorrect?

In general, I don't think it hurts to have the segment store throwing a more precise exception because we never know whether we will need to throw something different in the future for that same call.

Reply:
The Segment has been inactive for a while, so the segment store is evicting it from memory, hence closing its read index and cancelling all pending future reads. A Pending Future read does not qualify as "keeping it active" because, if that were the case, it would keep most segments active for no reason.

What is so imprecise about `CancellationException`? It means the future read has been cancelled due to internal system triggers. It does not mean it was cancelled due to an error. The same exception is thrown when the container is shutting down.

IMO, the behavior in this case is to go back to the client and figure out if the read is still needed. If so, then simply reinvoke it (from the client), as the segment may now live on another host.

Reply:
@andreipaduroiu I think you are clarifying that cancellation exception only happens due to a timeout. Perhaps because it is an unchecked exception, it isn't documented in the javadoc, but the user of the API could incorrectly think that it is being cancelled due to an error. If it is not too inconvenient, I'd rather resolve the ambiguity by qualifying the cancellation exception.

If changing the exception raised isn't a good path, then at least clarify in the javadoc what happens when a timeout occurs, like @tkaitchuck suggested. Currently, the javadoc says:

```
* @param timeout           Timeout for the operation.
```

but it doesn't say what happens as a consequence of the timeout. From experience, we learned that it throws `CancellationException`, but it was still not clear if that's the only case until @andreipaduroiu chimed in.

Reply:
Cancellation Exception happens due to the segment getting evicted or the container shutting down. Timeouts are not enforced in the segment store.

FYI, this is a CompletableFuture that's being waited upon. It should totally be expected that a Future be cancelled.

Reply:
Ok, so what I'm reading is that the handling of `CancellationException` should be uniform, and I think the recommendation so far is to return an empty `SegmentRead` result. But, I'm wondering why we can't simply be honest to the client and say that the request was cancelled. The client can decide whether to issue another read, throw an error up the call path, or something else. I'm not convinced that it is a good idea to mask it from the reader.

If this observation makes sense, then the follow-up question is how a client should treat a cancelled read request, is it always retryable?

Reply:
Updated the java doc for `StreamSegmentStore.read()` and created issue https://github.com/pravega/pravega/issues/2748 for handling CancellationException due to Segment container shutdown.